### DISCUSSION TITLE
ADC1?

### DISCUSSION THREAD

**Discussion by mechatroniks-git** [SpenceKonde/megaTinyCore#684] [Q&A] [answered]
I see the init_ADC1(), and "The core does not provide an equivalent to analogRead() for ADC1 at this time."

Anyone have an example of how to read ADC1?

**Comment by SpenceKonde** [comment, marked as answer]
Well, you would probably start with analogRead().... 


```
  int analogRead(uint8_t pin) {

    check_valid_analog_pin(pin); //This would have to go because it has a different set of pins
    if (pin < 0x80) {
      // If high bit set, it's a channel, otherwise it's a digital pin so we look it up..
      pin = digitalPinToAnalogInput(pin); //Uhoh! We can't use a pin number here, because it will look up the pin for ADC0
    }

// this check isn't very important
    #if (PROGMEM_SIZE > 4096)
      // don't waste flash on smallest parts.
      if ((pin & 0x7F) > 0x1F) { // highest valid mux value for any 0 or 1-series part.
        return ADC_ERROR_BAD_PIN_OR_CHANNEL;
      }
    #endif
// End of stuff you should cut out.... 
    if (!ADC0.CTRLA & 0x01) return ADC_ERROR_DISABLED;
    pin &= 0x1F;
    /* Reference should be already set up */
    /* Select channel */
    ADC0.MUXPOS = (pin << ADC_MUXPOS_gp);

    #if defined(STRICT_ERROR_CHECKING)
      if (ADC0.COMMAND) return ADC_ERROR_BUSY;
    #endif

    /* Start conversion */
    ADC0.COMMAND = ADC_STCONV_bm;

    /* Wait for result ready */
    while (!(ADC0.INTFLAGS & ADC_RESRDY_bm));

    /* Combine two bytes */
    return ADC0.RES;
  }

```

calling init_ADC1() in setup, and changing every instance of ADC0 to ADC1, should get you there except for setting the reference. 

But that's as easy as 
``` c++
VREF.CTRLC=reference; 
/* where reference is one of these:
    VREF_ADC1REFSEL_0V55_gc
    VREF_ADC1REFSEL_1V1_gc
    VREF_ADC1REFSEL_2V5_gc
    VREF_ADC1REFSEL_4V34_gc
    VREF_ADC1REFSEL_1V5_gc
*/ if using the DACREF for AC1, then you need to instead do something like
VREF.CTRLC = (VREF_DAC1REFSEL_gm & VREF.CTRLC) | VREF_ADC1REFSEL_2V5_gc

// the only complication is switching between external VREF, internal reference, and VDD being in a different register for no good reason:

ADC1.CTRLC = (ADC1.CTRLC & (~ADC_REFSEL_gm)) | source
/* where source is one of these
    ADC_REFSEL_INTREF_gc
    ADC_REFSEL_VDDREF_gc
    ADC_REFSEL_VREFA_gc
*/
```

As for the matter of the pin number, I would just use the channel number - ADC_CH(num) simply OR's the value with 128, but it's a lot clearer to say analogRead(ADC_CH(5)) than it would be to say analogRead(0x80 | 5) or (worst of all) analogRead1(133)

The table pins to inputs is in I/O Multiplexing Considerations chapter of the datasheet.

**Reply by SpenceKonde** [reply]
The main reason why I don't include stuff that uses ADC1 is that it's rarely of much value - these parts aren't short of ADC pins, so usually if you're using ADC1, you're doing something wacky with it where anything we did would just get in the way (like throwing it into free run mode with the window comparator enabled, or somesuch). 

This all assumes you have a 1-series part with at least 16k of flash. 
Nothing else has the bonus ADC. 

I once put both ADC's into freerun mode on different pins, and was able to get the rest of the code tight enough that I could look up each ADC reading in a table of 1024 values, blast it out over SPI and be ready to grab the value from the other ADC and send it over SPI  (the destination being a dual DAC IC - it would read two analog values and blast them out to a cheap DAC module at the speed of the ADC, and I think I even cranked up the clock speed of the ADC, since the inputs were smooth,... And that was without even using any assembly!

**Comment by mechatroniks-git** [comment]
Yup, was using ATtiny1614, switched to 1616 due to chip availability at the time, and the lure of a smaller package. 

Then to make for an easy route, I picked pin_PC0. I'll fix it on the next rev and avoid ADC1.

**Reply by SpenceKonde** [reply]
aAAAHA! I understand now. I will note that as I said above, it's not that hard if you're not doing wonky stuff with the ADC. The main annoying thing is conversion of ADC channels and digital pin numbers imop

You've heard my theory on why they implemented that right? As a substitute for a proper differential ADC. The DX got effectively the same solution, only automated - they don't have a true differential ADC (the voltage must be lower than the reference voltage) , it's not until the 2-series and EA series that we get a real differential ADC! So include two identical ADCs, trigger from same event, and bada-bing.

**Reply by mechatroniks-git** [reply]
That's interesting, they were rolling these things out, while perfecting along the way I guess.  Hopefully there is no 3-series coming out.

Thanks for your help.

**Reply by SpenceKonde** [reply]
I woud love to see a 3 series because it gives us the preview of the next generation 2 years or so early. 

But I would bet a good deal of money that there isn't going to be on.e for many reasons.

**Comment by mechatroniks-git** [comment]
So it appears the latest core supports ADC1 now?

**Reply by SpenceKonde** [reply]
Yeah I just went ahead and added the functions. If you want to be boring and use it just like ADC0.
